### ISSUE TITLE
2.2.5 regression: Object reference not set to an instance of an object

### ISSUE THREAD

**Issue by JustArchi** [microsoft/testfx#895]
## Description
> **IMPORTANT**: if the defect is reproduced only in a workflow from within the Visual Studio IDE then do not report the issue here - instead, please report it using Visual Studio's "Send Feedback" option that can be accessed from the Help menu OR using this link https://developercommunity.visualstudio.com.
>
> **IMPORTANT**: Before posting an issue, please search the repository for similar ones. It might have been solved or reported already. 
> 
> For a defect specific to the MSTest V2 test framework, describe the issue you've observed.

Hello, one of my CIs when testing `TestAdapter` update from version `2.2.4` to `2.2.5` has failed (two times) with an error that looks like race-condition caused by `TestAdapter`. The same code/project has compiled fine within the same CI, I'm quite confident that this is a regression in `TestAdapter`. The test was done with `dotnet test` command.

## Steps to reproduce
> What steps can reproduce the defect?
> 
> Please share the setup, sample project, target platform (desktop, core, uap)

If my guess about race condition is correct, I suspect the combination of `[assembly: Parallelize(Scope = ExecutionScope.MethodLevel)]` in assembly info, and a `[DataTestMethod]` with two or more `[DataRow]` may trigger it, the chance may be higher with more tests/rows declared.

I'm unable to reproduce it on my own machine personally. The issue was reproduced within AppVeyor's CI, **[here](https://ci.appveyor.com/project/JustArchi/archisteamfarm/builds/39780616#L325)** and **[here](https://ci.appveyor.com/project/JustArchi/archisteamfarm/builds/39782325#L326)**, doesn't look like a coincidence or any environment fault, updating `TestAdapter` from `2.2.4` to `2.2.5` causes that regression. I've never seen this error before during the last several years of the development, and the commit introducing it was `TestAdapter` update: https://github.com/JustArchiNET/ArchiSteamFarm/commit/b7a8e367d56e19909884d413616a1ad524a961c4

If you feel lucky, you can try commit `b7a8e367d56e19909884d413616a1ad524a961c4` from `renovate/mstest.testadapter-2.x` branch of my ASF project: https://github.com/JustArchiNET/ArchiSteamFarm. Simply execute `dotnet test ArchiSteamFarm.Tests` within the repo, msbuild will do the rest. This is unfortunately the only lead I have, I couldn't reproduce it myself, neither in standalone repro, nor within my project. GitHub CI (actions) was also unable to trigger this, and I test with 3 different OSes there.

For reference, the full command triggering the issue was:

```
dotnet test ArchiSteamFarm.Tests -c "Release" -p:ContinuousIntegrationBuild=true -p:UseAppHost=false --nologo
```

Additional ref: https://github.com/JustArchiNET/ArchiSteamFarm/pull/2353

## Expected behavior
> Share the expected output

All of the existing, passing tests in `TestAdapter` in version `2.2.4` passing also in `2.2.5`.

https://ci.appveyor.com/project/JustArchi/archisteamfarm/builds/39782316#L303

## Actual behavior
> What is the behavior observed?

Tests failing with:

```
Test run for C:\projects\archisteamfarm\ArchiSteamFarm.Tests\bin\Release\net48\ArchiSteamFarm.Tests.dll (.NETFramework,Version=v4.8)
Starting test execution, please wait...
A total of 1 test files matched the specified pattern.
  Failed DoesntSplitInTheMiddleOfMultiByteChar (True)
  Error Message:
   Object reference not set to an instance of an object.
  Stack Trace:
     at System.Collections.Generic.Dictionary`2.Insert(TKey key, TValue value, Boolean add)
```

https://ci.appveyor.com/project/JustArchi/archisteamfarm/builds/39782325#L326

## Environment
> Please share additional details about the test environment.
>  - Operating system
>  - Build version of vstest.console
>  - Package version of MSTest framework and adapter
>  - Other installed packages and their versions on the test project

Setup:

```
.NET SDK (reflecting any global.json):
 Version:   5.0.301
 Commit:    ef17233f86
Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.17763
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.301\
Host (useful for support):
  Version: 5.0.7
  Commit:  556582d964
```

I'm testing `net5.0` and `net48` targets, the `net5.0` seems to work correctly, AppVeyor reports issue only with `net48`. Not sure if that's relevant, perhaps `net48` is simply slower or there is some other difference that could cause that.

All packages of the test project:

```csproj
    <PackageVersion Include="ConfigureAwaitChecker.Analyzer" Version="5.0.0" /> <!-- Analyzer only, not a runtime lib -->
    <PackageVersion Include="Microsoft.NET.Test.Sdk" Version="16.10.0" />
    <PackageVersion Include="MSTest.TestAdapter" Version="2.2.5" />
    <PackageVersion Include="MSTest.TestFramework" Version="2.2.5" />
```

Link to the affected repo (and branch): https://github.com/JustArchiNET/ArchiSteamFarm/tree/renovate/mstest.testadapter-2.x

## My own observations

Judging by the history of changes, I blindly guess that the culprit could be introduced in **[this commit](https://github.com/microsoft/testfx/commit/57d2d00637e2bb3b5b9c4f9334a10ef2f6463146#diff-b18e5e3827d610658ab2e353ebd20f6510eb6112866f99fe381e4ca266ff7acdR16)**, as `SerializerCache` dictionary was introduced, but the subsequent calls operating on the dictionary are not synchronized, neither the dictionary is concurrent. This is a total blind guess though.

I wish I could be of more help, but since I'm unable to come up with a repro, I nailed this one down to as much information as I could realistically provide. Perhaps you'll have more luck than me reproducing this issue.

Thank you in advance for taking a look into it.

**Comment by JustArchi** [comment]
Side note: as expected, the 3rd CI try actually succeeded, so this is most definitely some kind of race condition with the internal structures, see https://ci.appveyor.com/project/JustArchi/archisteamfarm/builds/39785367

**Comment by Haplois** [comment]
Hi @JustArchi, thank you for the feedback.

Thank you for all your investigations! If it's not possible or easy to create a repro; can you add `--diag:diag.log` to your test runner, and share the resulting `*.log` files with us? (It should at least create to files).

```console
dotnet test ArchiSteamFarm.Tests -c "$env:CONFIGURATION" -p:ContinuousIntegrationBuild=true -p:UseAppHost=false --nologo --diag:diag.log
```

**Comment by JustArchi** [comment]
Hey @Haplois, thanks for answer.

> I was checking out your repo (see), we recommend keeping MSTest.TestAdapter and MSTest.TestFramework package versions same. It wont cause any problems between 2.2.4 and 2.2.5; but it might cause weird and hard to diagnose issues.

This is because my renovate bot responsible for dependencies managed to bump the Framework to 2.2.5 without issues but failed at Adapter, I have them synchronized normally. Regardless, the bug is reproduced with framework and adapter both at being `2.2.5`.

> Thank you for all your investigations! If it's not possible or easy to create a repro; can you add --diag:diag.log to your test runner, and share the resulting *.log files with us? (It should at least create to files).

I'm on it, might require a few tries to reproduce though: https://github.com/JustArchiNET/ArchiSteamFarm/commit/472acb03ea5ec668c6e6c05c9b3b51bc4bbea512

I'll post a reply once I manage to grab some logs, shouldn't take too long.

**Comment by JustArchi** [comment]
@Haplois got it first try. I'm attaching the full log from my CI run, it's not easy to grab the files from failed runs so I just configured my CI to print them on the console upon failure right before returning.

[log.txt](https://github.com/microsoft/testfx/files/6732818/log.txt)

The diag logs start at line 323 and continue all the way to the end of the file. I used `type *.log`, so there should be all generated.

Hope it helps!

**Comment by abatishchev** [comment]
I experienced the same. deleted the `.vs` folder and the error was gone.

My configuration:

```xml
<PackageReference Include="Microsoft.NET.Test.Sdk" Version="16.10.0" />
<PackageReference Include="MSTest.TestAdapter" Version="2.2.5" />
<PackageReference Include="MSTest.TestFramework" Version="2.2.5" />
```

Not saying this is not an issue, it is. But at least this workaround worked for me.

**Comment by JustArchi** [comment]
I don't have a `.vs` folder in my project and `dotnet test` probably doesn't create one, does it? It could be a total coincidence, unless I'm missing something. I have an impression that only Visual Studio IDE creates that one.

**Comment by abatishchev** [comment]
yes, you're right, @JustAchi! `.vs` folder is an artifact of VS and your environment is different. But I can swear I got the same exception on the very first run. Then I got much more serious trouble (reported as another regression in 2.2.5).

**Comment by Haplois** [comment]
After investigating this, we found two reasons for this issue
 - Because of a bug (#869) introduced in 2.2.4, if your tests contains any data which is not serializable - you might get this error. Workaround is either upgrading to 2.2.5+, or downgrading to 2.2.3. After version change, you might need to clean the solution, and rebuild it.
 
 - It might be because of a race condition (#906) in out internal serializer. Current workaround is downgrading to 2.2.3. You can update first version after #906 is fixed.

Fix for this issue is tracked in #906, closing this issue.
